### PR TITLE
Internal: Add Sticky upgrade promotion in widget panel [ED-13720]

### PR DESCRIPTION
**Pull request by elchugreeva**

### CODE REVIEW

File: assets/dev/scss/editor/panel/_categories.scss [hunk 1]
```diff
@@ -17,6 +33,12 @@
 		}
 	}
 
+	.elementor-upgrade-sticky {
+		position: sticky;
+		bottom:0;
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/_categories.scss:38]:
```suggestion
		bottom: 0;
```

File: assets/dev/scss/editor/panel/_categories.scss [hunk 2]
```diff
@@ -17,6 +33,12 @@
...
+
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/_categories.scss:39]: Remove empty line.

File: assets/dev/scss/editor/panel/panel.scss [hunk 3]
```diff
@@ -39,6 +39,11 @@
 		> .ui-resizable-w {
 			display: none;
 		}
+	#elementor-panel-elements-wrapper {
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/panel.scss:42]: Add empty line above.

File: assets/dev/scss/editor/panel/panel.scss [hunk 4]
```diff
@@ -39,6 +39,11 @@
...
+		flex: 1;
+		display: flex;
+		flex-direction: column;
+		}
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/panel.scss:46]: Remove?

```suggestion
```

File: assets/dev/scss/editor/panel/panel.scss [hunk 5]
```diff
@@ -114,6 +119,29 @@
 	}
 }
 
+#elementor-panel-get-pro-elements-sticky {
+	position: sticky;
+	bottom:0;
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/panel.scss:124]:
```suggestion
	bottom: 0;
```

File: assets/dev/scss/editor/panel/panel.scss [hunk 6]
```diff
@@ -114,6 +119,29 @@
...
+	font-size: 13px;
+	font-weight: 500;
+
+	.elementor-get-pro-sticky-message {
+		height: 40px;
+		padding: 9px 35px;
+		gap: 10px;
+		background-color: #FFFFFF;
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/panel.scss:132]: You will need to use css variable for the colors, to accommodate the light and dark modes.

File: includes/editor-templates/panel-elements.php [hunk 7]
```diff
@@ -12,25 +12,45 @@
...
 	<div id="elementor-panel-categories"></div>
-
-	<?php
-	$get_pro_details = apply_filters( 'elementor/editor/panel/get_pro_details', [
-		'link' => 'https://go.elementor.com/pro-widgets/',
-		'message' => __( 'Get more with Elementor Pro', 'elementor' ),
-		'button_text' => __( 'Upgrade Now', 'elementor' ),
-	] );
-	?>
+			<?php
```
Comment (hein-obox) [inline includes/editor-templates/panel-elements.php:24]: Do have to make changes here? I would try to keep the number of changes as low as possible, to make the review process easier.

File: includes/editor-templates/panel-elements.php [hunk 8]
```diff
@@ -12,25 +12,45 @@
...
+
+			<?php
+			$get_pro_details_sticky = apply_filters( 'elementor/editor/panel/get_pro_details', [
+					'link' => 'https://go.elementor.com/go-pro-sticky-widget-panel/',
+					'message' => __( 'Access all Pro widgets.', 'elementor' ),
+					'button_text' => __( 'Upgrade Now', 'elementor' ),
+			] );
+			?>
+			<div id="elementor-panel-get-pro-elements-sticky">
+				<img class="elementor-nerd-box-icon" src="<?php echo ELEMENTOR_ASSETS_URL . 'images/unlock
```
Comment (hein-obox) [inline includes/editor-templates/panel-elements.php:46]: Remove whitespace between image name and extension:

```suggestion
				<img class="elementor-nerd-box-icon" src="<?php echo ELEMENTOR_ASSETS_URL . 'images/unlock.svg';
```

File: assets/dev/scss/editor/panel/_categories.scss [hunk 9]
```diff
@@ -17,6 +33,12 @@
 		}
 	}
 
+	.elementor-upgrade-sticky {
+		position: sticky;
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/_categories.scss:37]: Maybe?

```suggestion
		position: fixed;
```

File: assets/dev/scss/editor/panel/panel.scss [hunk 10]
```diff
@@ -39,6 +39,11 @@
 		> .ui-resizable-w {
 			display: none;
 		}
+	#elementor-panel-elements-wrapper {
+		flex: 1;
```
Comment (hein-obox) [inline assets/dev/scss/editor/panel/panel.scss:43]: Necessary?

File: includes/editor-templates/panel-elements.php [hunk 11]
```diff
@@ -83,3 +94,4 @@
 		<a class="elementor-button go-pro" target="_blank" href="https://go.elementor.com/pro-global/"><?php echo esc_html__( 'Upgrade Now', 'elementor' ); ?></a>
 	</div>
 </script>
+
```
Comment (louiswol94) [inline includes/editor-templates/panel-elements.php:97]:
```suggestion
```

File: assets/dev/scss/editor/panel/panel.scss [hunk 12]
```diff
@@ -114,11 +120,44 @@
...
+	.elementor-get-pro-sticky-message {
+		margin-top: -14px;
+		min-height: 40px;
+		display: block;
+		padding: 9px min(35px, 5%);
+		gap: 10px;
+		background-color: var(--e-a-color-white);
+		color: var(--e-a-color-black);
+		text-align: center;
+		box-shadow: 0px -10px 14px -6px rgba(0, 0, 0, 0.05);
```
Comment (louiswol94) [inline assets/dev/scss/editor/panel/panel.scss:141]: `box-shadow: 0px -10px 14px -6px rgba(0, 0, 0, 0.05);`

Does this work well in RTL?

**Review by louiswol94 — CHANGES_REQUESTED** [review]
Is there a way to get the screenshot to not have so much white space at the bottom?

File: includes/editor-templates/panel-elements.php [hunk 13]
```diff
@@ -17,20 +17,27 @@
 
 <script type="text/template" id="tmpl-elementor-panel-categories">
 	<div id="elementor-panel-categories"></div>
-
 	<?php
 	$get_pro_details = apply_filters( 'elementor/editor/panel/get_pro_details', [
```
Comment (hein-obox) [inline includes/editor-templates/panel-elements.php:21]: Currently I am seeing this panel inside the Pro version as well.

File: assets/dev/scss/editor/panel/panel.scss (hunk 12 above)
Comment (hein-obox) [reply assets/dev/scss/editor/panel/panel.scss:140]: @louiswol94 I have tested it visually with RTL. I think it looks fine.

File: includes/editor-templates/panel-elements.php [hunk 14]
```diff
@@ -17,20 +17,27 @@
...
 	?>
-
 	<div id="elementor-panel-get-pro-elements" class="elementor-nerd-box">
 		<img class="elementor-nerd-box-icon" src="<?php echo ELEMENTOR_ASSETS_URL . 'images/go-pro.svg'; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>" loading="lazy" alt="<?php echo esc_attr__( 'Upgrade', 'elementor' ); ?>" />
 		<div class="elementor-nerd-box-message"><?php echo esc_html( $get_pro_details['message'] ); ?></div>
 		<a class="elementor-button go-pro" target="_blank" href="<?php echo esc_url( $get_pro_details['link'] ); ?>"><?php echo esc_html( $get_pro_details['button_text'] ); ?></a>
 	</div>
+	<div id="elementor-panel-get-pro-elements-sticky">
+		<img class="elementor-nerd-box-icon" src="<?php echo ELEMENTOR_ASSETS_URL . 'images/unlock-sticky.svg'; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>" loading="lazy" alt="<?php echo esc_attr__( 'Upgrade', 'elementor' ); ?>"/>
+		<div class="elementor-get-pro-sticky-message"><?php echo esc_html( $get_pro_details_sticky['message'] ); ?><a class="" target="_blank" href="<?php echo esc_url( $get_pro_details_sticky['link'] ); ?>"><?php echo esc_html( $get_pro_details_sticky['button_text'] ); ?></a></div>
```
Comment (hein-obox) [inline includes/editor-templates/panel-elements.php:39]: I don't think that we should add an empty class attribute.

```suggestion
		<div class="elementor-get-pro-sticky-message"><?php echo esc_html( $get_pro_details_sticky['message'] ); ?><a target="_blank" href="<?php echo esc_url( $get_pro_details_sticky['link'] ); ?>"><?php echo esc_html( $get_pro_details_sticky['button_text'] ); ?></a></div>
```

File: includes/editor-templates/panel-elements.php [hunk 15]
```diff
@@ -17,20 +17,31 @@
...
 		'message' => __( 'Get more with Elementor Pro', 'elementor' ),
 		'button_text' => __( 'Upgrade Now', 'elementor' ),
 	] );
 	?>
+	<?php
+	$get_pro_details_sticky = apply_filters( 'elementor/editor/panel/get_pro_details-sticky', [
+		'link' => 'https://go.elementor.com/go-pro-sticky-widget-panel/',
+		'message' => __( 'Access all Pro widgets. ', 'elementor' ),
+		'button_text' => __( 'Upgrade Now', 'elementor' ),
+	] );
```
[the remark this answers is not in the document]
Comment (hein-obox) [reply includes/editor-templates/panel-elements.php:30]: @elchugreeva @davseve Can you please check this? I am not sure what is expected. I would have expected `$get_pro_details` to have been run through the filter as well.

[the remark this answers is not in the document]
Comment (louiswol94) [reply includes/editor-templates/panel-elements.php:30]: @elchugreeva `Filtered_Promotions_Manager::get_filtered_promotion_data()` ?

File: tests/playwright/sanity/modules/promotions/promotions.test.ts [hunk 16]
```diff
@@ -102,4 +102,42 @@ test.describe( 'Promotion tests @promotions', () => {
...
+			// Act.
+			await wpAdminPage.openNewPage();
+			const parentContainer = page.locator( wrapperContainer );
+			const promoContainer = page.locator( promotionContainer );
+			await promoContainer.waitFor();
+
+			// Assert.
+			await expect( parentContainer ).toHaveScreenshot( `go-pro-sticky.png` );
+		} );
+		await test.step( 'Sticky Free to Pro - Top Bar On', async () => {
```
Comment (dmaziar) [inline tests/playwright/sanity/modules/promotions/promotions.test.ts:125]: this is two separate tests. Please split it

File: includes/editor-templates/panel-elements.php (hunk 13 above)
Comment (davseve) [reply includes/editor-templates/panel-elements.php:21]: @ha
AFAIK, You should see this promotion when Pro is activated.
In Pro it is filtered by a different filter. 
The `free to pro` promotion leads you to `https://elementor.com/pro/?utm_source=panel-widgets&utm_campaign=gopro&utm_medium=wp-dash`3
While in pro, the link leads to `wp-admin/admin.php?page=elementor-apps`

Furthermore, you can see that the promotion text in free and in Pro is different.

File: assets/dev/scss/editor/panel/panel.scss [hunk 17]
```diff
@@ -114,11 +120,44 @@
...
+		margin-left: 17px;
+	}
+}
+
 #elementor-panel-notice-wrapper .elementor-panel-notice {
 	width: 90%;
 	margin: 0 auto;
 	font-style: unset;
 	margin-block-end: 15px;
+
```
Comment (ronenelementor) [inline assets/dev/scss/editor/panel/panel.scss:160]: I think that this is also unnecessary according to Hein

File: assets/dev/scss/editor/panel/_categories.scss (hunk 1 above)
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/_categories.scss]: fixed

File: assets/dev/scss/editor/panel/_categories.scss (hunk 2 above)
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/_categories.scss]: fixed

File: assets/dev/scss/editor/panel/panel.scss (hunk 3 above)
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/panel.scss]: fixed

File: assets/dev/scss/editor/panel/panel.scss (hunk 5 above)
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/panel.scss]: fixed

File: assets/dev/scss/editor/panel/panel.scss [hunk 18]
```diff
@@ -114,6 +119,29 @@
...
+	font-weight: 500;
+
+	.elementor-get-pro-sticky-message {
+		height: 40px;
+		padding: 9px 35px;
+		gap: 10px;
+		background-color: #FFFFFF;
+		text-align: center;
+		margin-top: -9px;
+
```
[the remark this answers is not in the document]
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/panel.scss:141]: fixed

File: assets/dev/scss/editor/panel/panel.scss (hunk 6 above)
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/panel.scss]: fixed

File: includes/editor-templates/panel-elements.php (hunk 8 above)
Comment (elchugreeva) [reply includes/editor-templates/panel-elements.php]: fixed

File: assets/dev/scss/editor/panel/_categories.scss (hunk 9 above)
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/_categories.scss]: fixed

File: assets/dev/scss/editor/panel/panel.scss (hunk 10 above)
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/panel.scss]: fixed

File: includes/editor-templates/panel-elements.php [hunk 19]
```diff
@@ -17,20 +17,31 @@
...
 	<div id="elementor-panel-categories"></div>
-
 	<?php
 	$get_pro_details = apply_filters( 'elementor/editor/panel/get_pro_details', [
 		'link' => 'https://go.elementor.com/pro-widgets/',
 		'message' => __( 'Get more with Elementor Pro', 'elementor' ),
 		'button_text' => __( 'Upgrade Now', 'elementor' ),
 	] );
 	?>
+	<?php
```
[the remark this answers is not in the document]
Comment (elchugreeva) [reply includes/editor-templates/panel-elements.php]: fixed

File: includes/editor-templates/panel-elements.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (elchugreeva) [reply includes/editor-templates/panel-elements.php]: fixed

File: includes/editor-templates/panel-elements.php [hunk 20]
```diff
@@ -17,20 +17,31 @@
...
 	<div id="elementor-panel-get-pro-elements" class="elementor-nerd-box">
 		<img class="elementor-nerd-box-icon" src="<?php echo ELEMENTOR_ASSETS_URL . 'images/go-pro.svg'; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>" loading="lazy" alt="<?php echo esc_attr__( 'Upgrade', 'elementor' ); ?>" />
 		<div class="elementor-nerd-box-message"><?php echo esc_html( $get_pro_details['message'] ); ?></div>
 		<a class="elementor-button go-pro" target="_blank" href="<?php echo esc_url( $get_pro_details['link'] ); ?>"><?php echo esc_html( $get_pro_details['button_text'] ); ?></a>
 	</div>
+	<div id="elementor-panel-get-pro-elements-sticky">
+		<img class="elementor-nerd-box-icon" src="<?php echo ELEMENTOR_ASSETS_URL . 'images/unlock-sticky.svg'; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>" loading="lazy" alt="<?php echo esc_attr__( 'Upgrade', 'elementor' ); ?>"/>
+		<div class="elementor-get-pro-sticky-message"><?php echo esc_html( $get_pro_details_sticky['message'] ); ?><a class="" target="_blank" href="<?php echo esc_url( $get_pro_details_sticky['link'] ); ?>"><?php echo esc_html( $get_pro_details_sticky['button_text'] ); ?></a></div>
+	</div>
+
```
[the remark this answers is not in the document]
Comment (elchugreeva) [reply includes/editor-templates/panel-elements.php]: fixed

File: includes/editor-templates/panel-elements.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (davseve) [reply includes/editor-templates/panel-elements.php]: > $get_pro_details
@hein-obox 
This filter is not at the scope of this task.

File: includes/editor-templates/panel-elements.php [hunk 21]
```diff
@@ -17,20 +20,32 @@
...
 		'message' => __( 'Get more with Elementor Pro', 'elementor' ),
 		'button_text' => __( 'Upgrade Now', 'elementor' ),
-	] );
+	];
+	$promotion_data_sticky = [
+		'url' => 'https://go.elementor.com/go-pro-sticky-widget-panel/',
+		'message' => __( 'Access all Pro widgets. ', 'elementor' ),
+		'button_text' => __( 'Upgrade Now', 'elementor' ),
+	];
+	$get_pro_details = Filtered_Promotions_Manager::get_filtered_promotion_data( $get_pro_details, 'elementor/editor/panel/get_pro_details', 'link' );
```
Comment (louiswol94) [inline includes/editor-templates/panel-elements.php:34]: `$get_pro_details `- this one cannot go through the filter. Only put yours through a filter.

File: tests/playwright/sanity/modules/promotions/promotions.test.ts [hunk 22]
```diff
@@ -102,4 +102,42 @@ test.describe( 'Promotion tests @promotions', () => {
...
+			await expect( parentContainer ).toHaveScreenshot( `go-pro-sticky.png` );
+		} );
+
+	test( 'Promotions - Sticky Free to Pro - Top Bar On', async ( { page }, testInfo ) => {
+			// Arrange.
+			const wpAdminPage = new WpAdminPage( page, testInfo );
+			await wpAdminPage.setExperiments( {
+				editor_v2: true,
+			} );
+			const wrapperContainer = '#elementor-panel-categories',
```
Comment (hein-obox) [inline tests/playwright/sanity/modules/promotions/promotions.test.ts:131]: Same as above.

File: assets/dev/scss/editor/panel/panel.scss [hunk 23]
```diff
@@ -60,23 +65,23 @@
...
+  // Categories
+  @import "categories";
 
-	// Menu
-	@import "menu";
+  // Menu
+  @import "menu";
 
-	// Document Settings
+  // Document Settings
```
[the remark this answers is not in the document]
Comment (elchugreeva) [reply assets/dev/scss/editor/panel/panel.scss:77]: fixed

**Review by hein-obox — APPROVED** [review]

**Review by louiswol94 — APPROVED** [review]

**Pull request merged**